snapcraft/internal/build_providers/_multipass/_multipass.py

```diff
@@ -84,8 +88,67 @@ def _get_disk_image(self) -> str:
 
         return image
 
+    def _is_multipass_confined(self) -> bool:
+        if sys.platform != "linux":
+            return False
+
+        # Get the exit code of echo "ls $HOME/.local/share/snapcraft" | snap run --shell multipass;
+        # If 2, multipass *is* confined. Failure or other return codes indicate that either
+        # multipass or snap are not installed (we should not reach here then), or snapped multipass
+        # is installed but not confined.
+        in_snap_cmd = "ls '{}'\n".format(BaseDirectory.save_data_path("snapcraft"))
+        command = ["snap", "run", "--shell", "multipass"]
+        logger.debug('Running echo "{}" | {}'.format(in_snap_cmd, " ".join(command)))
+        try:
+            if (
+                subprocess.run(
+                    command,
+                    input=in_snap_cmd,
+                    encoding="utf-8",
+                    stdout=subprocess.DEVNULL,
+                    stderr=subprocess.DEVNULL,
+                ).returncode
+                == 2
+            ):
+                logger.debug("Multipass confined")
+                return True
+        except Exception:
+            pass
+
+        logger.debug("Multipass not confined")
+        return False
+
+    def _prepare_cloud_user_data(self) -> str:
+        # If Multipass Snap confined, place cloud-init file in location both snapcraft
+        # (classicly confined) and multipass can access. For Multipass, the host directory
+        # $HOME/snap/multipass/current corresponds to its confined $HOME
+        if self._is_multipass_confined():
```

Comment:
ooh, we should talk to the snapd team on a better way to support this. Maybe all this is not necessary by adding dotfile interface to multipass so it can read into snapcraft's internal directory structure?

Reply:
Hey, thanks for the rapid feedback!
Yes this code is all a workaround for the lack of such an interface. Definitely worth a chat with snapd team on the topic. At the very least, we need some way for 2 snaps to share a file.